### DISCUSSION TITLE
Stability matrix inversion

### DISCUSSION THREAD

**Discussion by OwlyWizard** [kc-ml2/meent#81] [General]
Hi, i am scratching my head over an error in meent,  I am trying to optimize a metasurface and I usually get the error " The algorithm failed to converge because the input matrix is ill-conditioned". I am not sure but I believe it happens more using vectorial shapes. I have seen the p_inv and perturbation options. I am not sure what this perturbation is about, is it like Tikhonov regularization or is it linked with the Enhanced transmission matrix approch discussed in your paper?

I suppose you get this error time to time, how do you proceed to solve it?
Right now I would proceed like this:
- try to increase the perturbation
- if it does not help, try to increase the number of modes (increase the simulation time)
- if it still does not help, use p_inv (increase a lot the simulation time)

Thank you in advance ~

**Comment by yonghakim** [comment]
Hi, please stop scratching your head :)

That is probably happening in eigen-decomposition. 
To avoid this, the matrix can be scanned before EIG to see if it's singular. If it is singular, we can give some perturbation to overcome. But as we know, this is an additional approximation. Please keep this in mind in your result analysis.

This functionality was implemented in the past but for some reason I deleted it.
I'll get back to you soon with the solution. 

BTW, can you share your experiment code that made trouble?

**Comment by yonghakim** [comment]
For each backend, there is `primitives.py` file in `emsolver` folder.
At the bottom, you can see `meeinv` function and please replace it with below code for avoiding singular matrix inversion by addding perturbation.


```python
def meeinv(x, use_pinv=False):

    try:
        if use_pinv:
            res = torch.linalg.pinv(x)
        else:
            res = torch.linalg.inv(x)

    except torch._C._LinAlgError as e:
        import logging
        logging.basicConfig(
            format='%(asctime)s; %(pathname)s; %(levelname)s:%(message)s',
            level=logging.WARNING,
            datefmt='%Y/%m/%d %H:%M:%S',
        )
        logging.warning('Matrix inversion failed in meeinv(). Adding perturbation.')
        x += torch.eye(len(x)) * 1E-20
        if use_pinv:
            res = torch.linalg.pinv(x)
        else:
            res = torch.linalg.inv(x)

    return res
```

**Comment by OwlyWizard** [comment]
Thank you a lot! I cannot share this code in particular (the seed of the optimization is a homemade metasurface) but I'll when I can (problem during a seedless optimization). Thank you again!

**Reply by yonghakim** [reply]
Never mind. The reason I asked the code is to make a working solution for your problem.
